rllib/connectors/tests/test_agent.py

```diff
@@ -381,44 +330,133 @@ def test_vr_connector_with_multiple_buffers(self):
         # This view requirement simulates the use-case of a decision transformer
         # without reward-to-go.
         view_rq_dict = {
-            # obs[t-context_len:t+1]
-            "context_obs": ViewRequirement("obs", shift=f"{-context_len}:1"),
-            # act[t-context_len:t]
-            "context_act": ViewRequirement("act", shift=f"{-context_len}:0"),
+            # obs[t-context_len+1:t]
+            "context_obs": ViewRequirement("obs", shift=f"-{context_len-1}:0"),
+            # next_obs[t-context_len+1:t]
+            "context_next_obs": ViewRequirement(
+                "obs", shift=f"-{context_len}:1", used_for_compute_actions=False
+            ),
+            # act[t-context_len+1:t]
+            "context_act": ViewRequirement(
+                SampleBatch.ACTIONS, shift=f"-{context_len-1}:-1"
+            ),
         }
 
         obs_arrs = np.arange(10)[:, None] + 1
-        act_arrs = np.arange(10)[:, None] * 100 + 1
+        act_arrs = (np.arange(10)[:, None] + 1) * 100
         n_steps = obs_arrs.shape[0]
-        ctx = ConnectorContext(view_requirements=view_rq_dict)
+        config = PPOConfig().to_dict()
+        ctx = ConnectorContext(
+            view_requirements=view_rq_dict, config=config, is_policy_recurrent=True
+        )
         c = ViewRequirementAgentConnector(ctx)
 
-        for is_training in [True, False]:
-            c.is_training(is_training)
-            for i in range(n_steps):
-                data = AgentConnectorDataType(
-                    0, 1, dict(obs=obs_arrs[i], act=act_arrs[i])
+        # keep a queue of length ctx_len of observations
+        obs_list, act_list = [], []
+        for t in range(n_steps):
+            # next state and action at time t-1 are the following
+            timestep_data = {
+                SampleBatch.NEXT_OBS: obs_arrs[t],
+                SampleBatch.ACTIONS: (
+                    np.zeros_like(act_arrs[0]) if t == 0 else act_arrs[t - 1]
+                ),
+                SampleBatch.T: t - 1,
+            }
+            data = AgentConnectorDataType(0, 1, timestep_data)
+            processed = c([data])
+            for_action = processed[0].data.for_action
+
+            if t == 0:
+                obs_list.extend([obs_arrs[0] for _ in range(context_len)])
+                act_list.extend(
+                    [np.zeros_like(act_arrs[0]) for _ in range(context_len)]
```

Comment:
just a nit picking TODO, maybe we should come back and make these actions different per time step.
don't know if that will reveal some deep bugs. right now everything is 0, things will match as long as the shapes match.
you don't need to add a TODO for this. if tests are already good.

Reply:
you need to do consistent zero padding for the learning part. If we don random padding it will mess up with the neural network that needs to learn that zero means invalid timestep.